doc/source/numpy_2_0_migration_guide.rst

```diff
@@ -420,12 +420,14 @@ The :ref:`copy keyword behavior changes <copy-keyword-changes-2.0>` in
    needed" in a way that's compatible with NumPy 1.x and 2.x, see
    `scipy#20172 <https://github.com/scipy/scipy/pull/20172>`__ for an example
    of how to do so.
-3. For any ``__array__`` method on a non-NumPy array-like object, a
-   ``copy=None`` keyword can be added to the signature - this will work with
-   older NumPy versions as well. If ``copy`` keyword is considered in
-   the ``__array__`` method implementation, then for ``copy=True`` always
-   return a new copy.
-
+3. For any ``__array__`` method on a non-NumPy array-like object,
+   ``dtype=None`` and ``copy=None`` keywords must be added to the signature -
+   this will work with older NumPy versions as well (although older numpy versions
+   will never pass in these keywords). If the keywords are actually used in the ``__array__``
+   method implementation, then for:
```

Comment:
```suggestion
   will never pass in these keywords). If the keywords are added to the ``__array__``
   signature, then for:
```

I still find "if the keywords are actually used in the implementation" a bit confusing, as it sounds that you are allowed to ignore them.

For `dtype` that is actually maybe true? (if the returned array doesn't match the requested `dtype`, I suppose that numpy will still do the cast?)

But once `copy` is _added to the signature_, it should always be honored in the implementation, right? (to ensure correct behaviour)

Reply:
Right, I agree that we should indicate that these arguments should be honored immediately (although `copy=True` is going to be passed to `__array__` since 2.1). Done!